tools/donate_cpu_lib.py

```diff
@@ -35,42 +37,58 @@ def check_requirements():
-            except:
-                print('Failed to clone, will try again in 10 minutes..')
-                time.sleep(600)
-                continue
-        time.sleep(2)
-        return True
-    if os.path.exists(cppcheck_path):
-        print('Failed to update Cppcheck sources, trying a fresh clone..')
+# Try and retry with exponential backoff if an exception is raised
+def try_retry(fun, fargs=(), max_tries=5):
+    sleep_duration = 5.0
+    for i in range(max_tries):
         try:
-            os.chdir(work_path)
-            shutil.rmtree(cppcheck_path)
-            get_cppcheck(cppcheck_path, work_path)
-        except:
-            print('Failed to remove Cppcheck folder, please manually remove ' + work_path)
-            return False
-    return False
+            return fun(*fargs)
+        except KeyboardInterrupt as e:
+            # Do not retry in case of user abort
+            raise e
+        except BaseException as e:
+            if i < max_tries - 1:
+                print("{} in {}: {}".format(type(e).__name__, fun.__name__, str(e)))
+                print("Trying {} again in {} seconds".format(fun.__name__, sleep_duration))
+                time.sleep(sleep_duration)
+                sleep_duration *= 2.0
+            else:
+                print("Maximum number of tries reached for {}".format(fun.__name__))
+                raise e
+
+
+def clone_cppcheck(repo_path):
+    repo_git_dir = os.path.join(repo_path, '.git')
+    if os.path.exists(repo_git_dir):
+        return
+    # A shallow git clone (depth = 1) is enough for building and scanning.
+    # Do not checkout until fetch_cppcheck_version.
+    subprocess.check_call(['git', 'clone', '--depth=1', '--no-checkout', CPPCHECK_REPO_URL, repo_path])
+    # Checkout an empty branch to allow "git worktree add" for main later on
+    os.chdir(repo_path)
+    try:
+        # git >= 2.27
+        subprocess.check_call(['git', 'switch', '--orphan', 'empty'])
+    except subprocess.CalledProcessError:
+        subprocess.check_call(['git', 'checkout','--orphan', 'empty'])
+
+
+def get_cppcheck_version(repo_path, version, cppcheck_path):
+    if os.path.exists(cppcheck_path):
+        os.chdir(cppcheck_path)
+        # It is possible to pull branches, not tags
+        if version == 'main':
+            subprocess.check_call(['git', 'checkout' , '-f', 'origin/' + version])
+            subprocess.check_call(['git', 'pull'])
+        else:
+            subprocess.check_call(['git', 'checkout' , '-f', version])
+    else:
+        os.chdir(repo_path)
+        if version != 'main':
+            # Since this is a shallow clone, explicitly fetch the remote version tag
+            refspec = 'refs/tags/' + version + ':ref/tags/' + version
+            subprocess.check_call(['git', 'fetch', '--depth=1', 'origin', refspec])
+        subprocess.check_call(['git', 'worktree', 'add', cppcheck_path,  version])
```

Comment:
Can this be set earlier so we don't need `os.chdir()` (maybe not at all)? I am not familiar with this at all.

Reply:
I'm not sure I understand the suggestion. 

"git worktree add" requires the working dir to be a (cloned) repo, so os.chdir() is needed. 
Earlier, we don't know which version we want a worktree for.
os.chdir() is a fairly lightweight operation so I don't see the worry.

Reply:
> I'm not sure I understand the suggestion.

I just don't understand "git worktree" yet ;)

> "git worktree add" requires the working dir to be a (cloned) repo, so os.chdir() is needed.
> Earlier, we don't know which version we want a worktree for.
> os.chdir() is a fairly lightweight operation so I don't see the worry.

I see. `chdir()` is just plain dangerous since you might lose your actual working directory so if possible you should avoid using it.

Reply:
That's a fair point, but in my defense I'm replicating existing chdir usage. 

Will try removing some chdir() calls and using subprocess' cwd parameter instead.

Reply:
> That's a fair point, but in my defense I'm replicating existing chdir usage.

That's fine. I just stumbled across `git worktree` before I saw you are actually using it and thought it might help - not knowing what it actually does.

> Will try removing some chdir() calls and using subprocess' cwd parameter instead.

I didn't think of that at all. Nice one!

Reply:
With typical git usage, the file tree (eg your source files) and git data (blobs and metadata under .git) are located in the same directory select during "git clone". 

git worktree allows placing the file tree into a completely separate directory (although they still keep a ref to each other). 
This also allow having multiple file trees, for instance when working with different branches. 

Without this, you'd have to either switch back and forth between branch, or do multiple clones which increase  storage+bandwidth usage.

Reply:
Thanks for the explanation. So I mainly got it right from reading the documentation. :)